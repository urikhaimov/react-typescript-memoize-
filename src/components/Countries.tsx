import React, { useEffect, useState } from 'react';
import useCountriesService from '../services/useCountriesService';
import Loader from './Loader';
import CountryCard from './CountryCard';
import { useLocalStorage } from 'usehooks-ts'
import { Country } from '../types/Country';
import _ from 'lodash';
import SearchBar from './SearchBar';

const Countries: React.FC<{}> = () => {
  const service = useCountriesService();
  const [countriesFromStorage, setCountriesFromStorage] = useLocalStorage('countries', new Array<Country>())
  const [countries, setCountries] = useState(new Array<Country>())
  const [allCountries, setAllCountries] = useState(new Array<Country>())

  useEffect(() => {
    if (service.status === 'loaded') {
      const updatedCountries = getUniqueListByName([...countriesFromStorage, ...service.payload.results])
      setCountries(updatedCountries);
      setAllCountries(updatedCountries)
    }
  }, [service, countriesFromStorage])

  const getUniqueListByName = (arr: Array<Country>) =>
    [...new Map(arr.map(item => [item.name.official, item])).values()]

  const isCountryExistsInStorage = (current: Object) =>
    countriesFromStorage.some(({ idd }) =>
      _.isEqual(idd, current))

  const addCountryToStorage = (country: Country) => {
    if (!isCountryExistsInStorage(country.idd)) {
      setCountriesFromStorage([country, ...countriesFromStorage])
    }

  }

  const filterCountries = (e: any) => {
    setCountries(allCountries.filter((c: Country) =>
      c.name.official.toLowerCase().includes(e.target.value.toLowerCase())
    ))
  }

  return (
    <>
      <div className="country-container ">
        <SearchBar
          changeHandler={filterCountries} />
        {service.status === 'loading' && (
          <div className="loader-container">
            <Loader />
          </div>
        )}
        {service.status === 'loaded' &&
          countries.map((country, i) => (
            <CountryCard
              key={`countryCard-${i}`}
              country={country}
              clickHandler={addCountryToStorage} />
          ))}

      </div>
      {service.status === 'error' && (
        <div>Error!</div>
      )}
    </>
  );
};

export default Countries;
