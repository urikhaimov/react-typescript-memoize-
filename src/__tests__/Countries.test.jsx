import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, fireEvent, render } from '@testing-library/react';
import Countries from '../components/Countries';
import request from './__mocks__/request';


afterEach(cleanup);
// The assertion for a promise must be returned.
it('works with async/await', async () => {
  expect.assertions(1);
  const country = await request();
  expect(country.length).toEqual(2)
});


test('render Countries', () => {

  const component = renderer.create(
    <Countries/>,

  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
}) 
