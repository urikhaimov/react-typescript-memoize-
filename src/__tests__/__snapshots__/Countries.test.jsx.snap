// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Countries 1`] = `
<div
  className="country-container "
>
  <div
    className="search-bar"
  >
    <label>
      Search
      <input
        id="searchbartext"
        onChange={[Function]}
      />
    </label>
  </div>
  <div
    className="loader-container"
  >
    <div
      className="loader"
    />
  </div>
</div>
`;
